File: 2024/day10/day10.py
import sys

grid = [line.strip() for line in sys.stdin]
trailheads = set()

for y, row in enumerate(grid):
    for x, cell in enumerate(row):
        if cell == '0':
            trailheads.add((y, x))

def out_of_bounds(y, x):
    return y < 0 or y >= len(grid) or x < 0 or x >= len(grid[0])

directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
part_1 = part_2 = 0

for y, x in trailheads:
    stack = [(y, x)]
    visited = set()
    reachable_nines = set()
    trailhead_rating = 0

    while stack:
        y, x = stack.pop()
        visited.add((y, x))

        if grid[y][x] == '9':
            reachable_nines.add((y, x))
            trailhead_rating += 1
            continue

        for dy, dx in directions:
            ny, nx = y + dy, x + dx
            if out_of_bounds(ny, nx):
                continue
            if ord(grid[ny][nx]) == ord(grid[y][x]) + 1:
                stack.append((ny, nx))

    part_1 += len(reachable_nines)
    part_2 += trailhead_rating

print(f'Part 1: {part_1}')
print(f'Part 2: {part_2}')

File: 2024/day10/day10.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct{ y, x int }

func main() {
	grid, trailheads := readInput()
	directions := []Point{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

	outOfBounds := func(p Point) bool {
		return p.y < 0 || p.x < 0 || p.y >= len(grid) || p.x >= len(grid[p.y])
	}

	part1, part2 := 0, 0
	for trailhead := range trailheads {
		stack := []Point{trailhead}
		visited := make(map[Point]bool)
		reachableNines := make(map[Point]bool)
		trailheadRating := 0

		for len(stack) > 0 {
			current := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			visited[current] = true

			if grid[current.y][current.x] == '9' {
				reachableNines[current] = true
				trailheadRating++
				continue
			}

			for _, dir := range directions {
				next := Point{current.y + dir.y, current.x + dir.x}
				if outOfBounds(next) {
					continue
				}
				if int(grid[next.y][next.x]) == int(grid[current.y][current.x])+1 {
					stack = append(stack, next)
				}
			}
		}

		part1 += len(reachableNines)
		part2 += trailheadRating
	}

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}

func readInput() ([]string, map[Point]bool) {
	scanner := bufio.NewScanner(os.Stdin)

	var grid []string
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	trailheads := make(map[Point]bool)
	for y, row := range grid {
		for x, cell := range row {
			if cell == '0' {
				trailheads[Point{y, x}] = true
			}
		}
	}

	return grid, trailheads
}
